Return EXIT_FAILURE on post-init fatal errors

furszy opened the pull request:
It seems odd to return `EXIT_SUCCESS` when the node aborted execution due a fatal internal error
or any post-init problem that triggers an unrequested shutdown.

e.g. blocks or coins db I/O errors, disconnect block failure, failure during thread import (external
blocks loading process error), among others.

furszy committed:
Return EXIT_FAILURE on post-init fatal errors

It seems odd to return `EXIT_SUCCESS` when the node aborted
execution due a fatal internal error or any post-init problem
that triggers an unrequested shutdown.

e.g. blocks or coins db I/O errors, disconnect block failure,
failure during thread import (external blocks loading process
error), among others.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

sedited reviewed (changes requested):
Was wondering if this was attempted before and skimmed through the pull request history, but I did not find prior attempts.

furszy commented on src/shutdown.h:
```diff
@@ -19,6 +19,9 @@ bool InitShutdownState();
 void StartShutdown();
 
+/** Request application shutdown due an internal error. */
+void StartErrorShutdown();
```
The reason behind that is https://github.com/bitcoin/bitcoin/pull/27607#discussion_r1199624849.
But.. could not expose it and move the caller to use `AbortNode()` instead. Same as we do with the external block import failures.

furszy replied:
Still, in order to do that in the cleanest possible way, will need to refactor `shutdown.cpp`, so `AbortNode` is placed after the static `StartErrorShutdown` function. Which I'm not so sure that worth it due the conflicts that could cause with your PRs. Thoughts?

TheCharlatan replied:
So IIRC, this call [here](https://github.com/bitcoin/bitcoin/commit/ca3041984cf3665e27b6783c923ab1c32682500a#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1669) would call `StartErrorShutdown`?

TheCharlatan replied:
> Which.. isn't bad as it will remove another ui_interface dependency.

Can you elaborate on this point? `AbortNode` still relies on `ui_interface` from what I can tell. 

From my current understanding both the `StartShutdown` in `ThreadImport` and in the check are call shutdown and node `AbortNode`, because `BlockImport` might run before the ui is available (this is me guessing out of historical reasons). I am not sold on this point though, since this runs in another thread and we also provide the noui consumer of the ui interface. So I think these should both become calls to `AbortNode`.

furszy replied:
> > Which.. isn't bad as it will remove another ui_interface dependency.
> 
> Can you elaborate on this point? `AbortNode` still relies on `ui_interface` from what I can tell.

Ok, I confused you. Sorry.

The ui_interface dependency that can be removed is the `base.cpp` one. Thus why said to "bubble up" the error string up to `init.cpp` instead of calling `InitError` from the index base class `Start` method.

> From my current understanding both the `StartShutdown` in `ThreadImport` and in the check are call shutdown and node `AbortNode`, because `BlockImport` might run before the ui is available (this is me guessing out of historical reasons).

`AbortNode` isn't doing anything special. We already use `InitError` in `init.cpp` prior spawning the import thread.
What the node will do when there is no gui is print the error to the command line, and the debug file, prior shutting down (using the noui consumer as you said).

> I am not sold on this point though, since this runs in another thread and we also provide the noui consumer of the ui interface. So I think these should both become calls to `AbortNode`.

Yeah, that is my idea as well. The `ThreadImport` "fatal" failures should use `AbortNode`.

furszy commented on src/shutdown.h:
```diff
@@ -29,7 +32,9 @@ bool ShutdownRequested();
 /** Wait for StartShutdown to be called in any thread. This can only be used
  * from a single thread.
+ * Returns true if shutdown was requested by the user. Otherwise, it's an
+ * internally requested shutdown which only happens when an error arises.
```
> This comment seems a bit inaccurate, e.g. we may have `stop_at_height` set, but shutdown is still triggered internally. There is also one error case in https://github.com/bitcoin/bitcoin/blob/master/src/node/blockstorage.cpp#L931 that triggers a normal shutdown.

I would label `stop_at_height` as an action requested by the user. The user had to provide `-stopatheight` at startup so the shutdown is expected.

On the other hand, https://github.com/bitcoin/bitcoin/blob/master/src/node/blockstorage.cpp#L931 seems to be a good candidate for `AbortNode`. It's a error in the `ThreadImport`, so either the db data got modified in some wrong way or the user provided invalid external blocks.

furszy commented on src/shutdown.h:
```diff
@@ -29,7 +32,9 @@ bool ShutdownRequested();
+ * internally requested shutdown which only happens when an error arises.
  */
-void WaitForShutdown();
+bool WaitForShutdown();
```
sure

furszy replied:
Ended up not adding this change to not modify `AbortShutdown()`.

furszy committed:
move ThreadImport ABC error to use AbortNode

'StartShutdown' should only be used for user requested
shutdowns. Internal errors that cause a shutdown should
use 'AbortNode'.

furszy commented:
Feedback tackled. Thanks TheCharlatan!. Changes:

* Moved ThreadImport ABC error to use `AbortNode`. Per [comment](https://github.com/bitcoin/bitcoin/pull/27708#discussion_r1206967261).
* And moved `AbortNode` code below `StartShutdown` so it can use the introduced static function instead of having to add a public `StartErrorShutdown` function that is only used by the .cpp file and not externally.

theStack reviewed (commented):
Concept ACK

Regarding the conclusion in the first commit (_"'StartShutdown' should only be used for user requested shutdowns. Internal errors that cause a shutdown should use 'AbortNode'."_), the following is probably another candidate where `AbortNode` is preferred over `StartShutdown`? 

https://github.com/bitcoin/bitcoin/blob/8cc65f093c0cf7a27b3bfc6da90fd7a6ac8f5e48/src/index/base.cpp#L33-L41

(not saying it necessarily has to happen in this PR though, if it makes sense).

furszy reviewed (commented):
> Regarding the conclusion in the first commit (_"'StartShutdown' should only be used for user requested shutdowns. Internal errors that cause a shutdown should use 'AbortNode'."_), the following is probably another candidate where `AbortNode` is preferred over `StartShutdown`?

Yeah absolutely. Good eye. If you compare it in-detail, that function is a plain copy of `AbortNode`: it (1) calls `SetMiscWarning`, (2) logs the error, (3) notifies the UI by calling `InitError` (noui listeners as well) and (4) calls to the shutdown function..
So, we can just get rid of it.

theStack committed:
refactor: index: use `AbortNode` in fatal error helper

Deduplicates code in the `FatalError` template function by using
`AbortNode` which does the exact same thing if called without any user
message (i.e. without second parameter specified). The template is still
kept for ease-of-use w.r.t. not having to call `tfm::format(...)` at the
call-side each time, and also to keep the diff minimal.

theStack commented:
> > Regarding the conclusion in the first commit (_"'StartShutdown' should only be used for user requested shutdowns. Internal errors that cause a shutdown should use 'AbortNode'."_), the following is probably another candidate where `AbortNode` is preferred over `StartShutdown`?
> 
> Yeah absolutely. Good eye. If you compare it in-detail, that function is a plain copy of `AbortNode`: it (1) calls `SetMiscWarning`, (2) logs the error, (3) notifies the UI by calling `InitError` (noui listeners as well) and (4) calls to the shutdown function.. So, we can just get rid of it.

Oh you're right, didn't look at `AbortNode` before, it's indeed almost the same. Feel free to pull in the dedup commit then: https://github.com/theStack/bitcoin/commit/3eca21000eb33b96063fdcdf30e8ad8514e41126 Note that I still kept the `FatalError` helper for ease-of-use for the callers, as it still is useful for not having to call `tfm::format(...)` each time (thanks to simple template magic).

furszy commented:
done, cherry-picked 🚜.

ryanofsky commented:
Approach ACK d44dd73bcabc694880dfc966fad21980cd3b4c64. Need to review more closely, but these changes look very good and should help with #27711

ryanofsky reviewed (approved):
Code review ACK d44dd73bcabc694880dfc966fad21980cd3b4c64. But it would be nice to extend this change to work with bitcoin-qt as well. I also think it would be good to apply the following changes, which replace the `bool shutdown_due_error` global with an `int exit_status{EXIT_SUCCESS}` `NodeContext` member. 

<details><summary>diff</summary>
<p>

```diff
diff --git a/src/bitcoind.cpp b/src/bitcoind.cpp
index c305d5e6cdc1..0a2a6ed4ebc2 100644
--- a/src/bitcoind.cpp
+++ b/src/bitcoind.cpp
@@ -169,6 +169,9 @@ static bool AppInit(NodeContext& node, int argc, char* argv[])
         // Set this early so that parameter interactions go to console
         InitLogging(args);
         InitParameterInteraction(args);
+        if (!InitShutdownState(node.exit_status)) {
+            return InitError(Untranslated("Initializing wait-for-shutdown state failed."));
+        }
         if (!AppInitBasicSetup(args)) {
             // InitError will have been called with detailed error, which ends up on console
             return false;
@@ -236,11 +239,6 @@ static bool AppInit(NodeContext& node, int argc, char* argv[])
     }
 #endif
     SetSyscallSandboxPolicy(SyscallSandboxPolicy::SHUTOFF);
-    if (fRet) {
-        fRet = WaitForShutdown();
-    }
-    Interrupt(node);
-    Shutdown(node);
 
     return fRet;
 }
@@ -264,5 +262,12 @@ MAIN_FUNCTION
     // Connect bitcoind signal handlers
     noui_connect();
 
-    return (AppInit(node, argc, argv) ? EXIT_SUCCESS : EXIT_FAILURE);
+    if (AppInit(node, argc, argv)) {
+        WaitForShutdown();
+    } else {
+        node.exit_status = EXIT_FAILURE;
+    }
+    Interrupt(node);
+    Shutdown(node);
+    return node.exit_status;
 }
diff --git a/src/init.cpp b/src/init.cpp
index 1ae21ec8d2dd..c4b09eb2373c 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -812,9 +812,6 @@ bool AppInitBasicSetup(const ArgsManager& args)
     // Enable heap terminate-on-corruption
     HeapSetInformation(nullptr, HeapEnableTerminationOnCorruption, nullptr, 0);
 #endif
-    if (!InitShutdownState()) {
-        return InitError(Untranslated("Initializing wait-for-shutdown state failed."));
-    }
 
     if (!SetupNetworking()) {
         return InitError(Untranslated("Initializing networking failed."));
diff --git a/src/node/context.h b/src/node/context.h
index 84f4053c8407..b24f778f4f64 100644
--- a/src/node/context.h
+++ b/src/node/context.h
@@ -7,7 +7,9 @@
 
 #include <kernel/context.h>
 
+#include <atomic>
 #include <cassert>
+#include <cstdlib>
 #include <functional>
 #include <memory>
 #include <vector>
@@ -62,6 +64,7 @@ struct NodeContext {
     interfaces::WalletLoader* wallet_loader{nullptr};
     std::unique_ptr<CScheduler> scheduler;
     std::function<void()> rpc_interruption_point = [] {};
+    std::atomic<int> exit_status{EXIT_SUCCESS};
 
     //! Declare default constructor and destructor that are not inline, so code
     //! instantiating the NodeContext struct doesn't need to #include class
diff --git a/src/shutdown.cpp b/src/shutdown.cpp
index 72f23be21ca0..7641ee3b066a 100644
--- a/src/shutdown.cpp
+++ b/src/shutdown.cpp
@@ -11,6 +11,7 @@
 
 #include <logging.h>
 #include <node/interface_ui.h>
+#include <util/check.h>
 #include <util/tokenpipe.h>
 #include <warnings.h>
 
@@ -21,8 +22,7 @@
 #endif
 
 static std::atomic<bool> fRequestShutdown(false);
-// Whether shutdown was requested by the user or because of an internal error
-static std::atomic<bool> shutdown_due_error{false};
+static std::atomic<int>* g_exit_status{nullptr};
 
 #ifdef WIN32
 /** On windows it is possible to simply use a condition variable. */
@@ -35,8 +35,9 @@ static TokenPipeEnd g_shutdown_r;
 static TokenPipeEnd g_shutdown_w;
 #endif
 
-bool InitShutdownState()
+bool InitShutdownState(std::atomic<int>& exit_status)
 {
+    g_exit_status = &exit_status;
 #ifndef WIN32
     std::optional<TokenPipe> pipe = TokenPipe::Make();
     if (!pipe) return false;
@@ -46,9 +47,8 @@ bool InitShutdownState()
     return true;
 }
 
-static void StartShutdown(bool error)
+void StartShutdown()
 {
-    if (error) shutdown_due_error.exchange(error);
 #ifdef WIN32
     std::unique_lock<std::mutex> lk(g_shutdown_mutex);
     fRequestShutdown = true;
@@ -68,8 +68,6 @@ static void StartShutdown(bool error)
 #endif
 }
 
-void StartShutdown() { return StartShutdown(/*error=*/false); }
-
 bool AbortNode(const std::string& strMessage, bilingual_str user_message)
 {
     SetMiscWarning(Untranslated(strMessage));
@@ -78,7 +76,8 @@ bool AbortNode(const std::string& strMessage, bilingual_str user_message)
         user_message = _("A fatal internal error occurred, see debug.log for details");
     }
     InitError(user_message);
-    StartShutdown(/*error=*/true);
+    Assert(g_exit_status)->store(EXIT_FAILURE);
+    StartShutdown();
     return false;
 }
 
@@ -97,7 +96,7 @@ bool ShutdownRequested()
     return fRequestShutdown;
 }
 
-bool WaitForShutdown()
+void WaitForShutdown()
 {
 #ifdef WIN32
     std::unique_lock<std::mutex> lk(g_shutdown_mutex);
@@ -109,5 +108,4 @@ bool WaitForShutdown()
         assert(0);
     }
 #endif
-    return !shutdown_due_error;
 }
diff --git a/src/shutdown.h b/src/shutdown.h
index 34d5d30bc074..2b8fdb779ac6 100644
--- a/src/shutdown.h
+++ b/src/shutdown.h
@@ -8,10 +8,12 @@
 
 #include <util/translation.h> // For bilingual_str
 
+#include <atomic>
+
 /** Initialize shutdown state. This must be called before using either StartShutdown(),
  * AbortShutdown() or WaitForShutdown(). Calling ShutdownRequested() is always safe.
  */
-bool InitShutdownState();
+bool InitShutdownState(std::atomic<int>& exit_status);
 
 /** Request shutdown of the application. */
 void StartShutdown();
@@ -29,9 +31,7 @@ bool ShutdownRequested();
 
 /** Wait for StartShutdown to be called in any thread. This can only be used
  * from a single thread.
- * Returns true if shutdown was requested by the user. Otherwise, it's an
- * internally requested shutdown which only happens when an error arises.
  */
-bool WaitForShutdown();
+void WaitForShutdown();
 
 #endif // BITCOIN_SHUTDOWN_H
```
</p>
</details>

I think this approach has a few advantages:

- Avoids adding a new shutdown global right before [#27711](https://github.com/bitcoin/bitcoin/pull/27711) removes all the shutdown globals, so there is less cleanup [#27711](https://github.com/bitcoin/bitcoin/pull/27711) needs to do later.
- I think semantics of an `exit_status` field are more obvious than a `shutdown_due_error` field. Easier to think concretely about what the exit code will be than abstractly about "whether shutdown was requested by the user".
- Having a `NodeContext::exit_status` field should make it easier to port this change to the GUI, since the GUI doesn't use the `WaitForShutdown` function.

(EDIT: Updated code suggestion above to fix early return bug in initial post)

furszy committed:
gui: return EXIT_FAILURE on post-init fatal errors

furszy reviewed (commented):
Thanks @ryanofsky! I see most of the changes great, except one.

If we change `AppInit` return value to be an int, we need to change all the
return lines to return an int, not only the ones at the end of the function.

Otherwise, the `return false` (or `return error(msg)`) statements will be treated as
`return 0` which means `EXIT_SUCCESS`, which is the opposite that we want.

Other than that nuance, pulling them :).

ryanofsky commented:
Oh, you are right. I thought I searched for early returns, but somehow I missed seeing them. I updated the suggestion above to fix this. I think AppInit should keep returning bool not int to keep the changes minimal, and because returning bool is better for following existing coding conventions.

furszy commented:
Thanks ryanofsky, updated per feedback with a tiny diff:

Didn't move `InitShutdownState` from `AppInitBasicSetup` to `AppInit` and instead provided the `NodeContext` ref to `AppInitBasicSetup`.
Not sure if you have an strong opinion about this (shoot if you have it) but the rationale was to keep the same workflow as we have now so the bitcoin-qt related changes b0267f2 doesn't introduce another `InitShutdownState` call.

> I think AppInit should keep returning bool not int to keep the changes minimal, and because returning bool is better for following existing coding conventions.

Funny enough, `GuiMain()` function returns an int. I think that we just don't have any convention here.
Still, agree to follow the simpler path.

ryanofsky reviewed (approved):
Code review ACK f13880e83200ec824d7528061f96708d96bf9bd4

This looks good as is, but would suggest two changes if you feel like making them:

1. Maybe drop the "move 'AbortNode' after 'StartShutdown'" commit if it is not needed anymore. Dropping it would help git blame work better and make history more legible. Also [#27711](https://github.com/bitcoin/bitcoin/pull/27711) is going to remove `AbortNode` so moving it now will cause an uglier conflict.

2. I think it would be better if the main() function always returned `node.exit_status` instead conditionally returning it. Better to make it clear that setting `node.exit_status` always determines the exit code and not create doubt about whether the value will be ignored. Would suggest the following (despite slightly longer diff)

```diff
diff --git a/src/bitcoind.cpp b/src/bitcoind.cpp
index 71b2eef814c6..4133269b3433 100644
--- a/src/bitcoind.cpp
+++ b/src/bitcoind.cpp
@@ -236,13 +236,6 @@ static bool AppInit(NodeContext& node, int argc, char* argv[])
     }
 #endif
     SetSyscallSandboxPolicy(SyscallSandboxPolicy::SHUTOFF);
-    if (fRet) {
-        WaitForShutdown();
-    } else {
-        node.exit_status = EXIT_FAILURE;
-    }
-    Interrupt(node);
-    Shutdown(node);
 
     return fRet;
 }
@@ -266,5 +259,13 @@ MAIN_FUNCTION
     // Connect bitcoind signal handlers
     noui_connect();
 
-    return AppInit(node, argc, argv) ? node.exit_status.load() : EXIT_FAILURE;
+    if (AppInit(node, argc, argv)) {
+        WaitForShutdown();
+    } else {
+        node.exit_status = EXIT_FAILURE;
+    }
+    Interrupt(node);
+    Shutdown(node);
+
+    return node.exit_status;
 }
```

I think this would be a general improvement anyway because it reduces complexity of `AppInit` and make it a pure initialization function, not a function that initializes, interrupts, and waits.

furszy reviewed (commented):
> 2. I think it would be better if the main() function always returned `node.exit_status` instead conditionally returning it. Better to make it clear that setting `node.exit_status` always determines the exit code and not create doubt about whether the value will be ignored. 
> I think this would be a general improvement anyway because it reduces complexity of `AppInit` and make it a pure initialization function, not a function that initializes, interrupts, and waits.

Hmm, isn't the diff exactly what I changed in my last push? check https://github.com/bitcoin/bitcoin/pull/27708#issuecomment-1583329813.

The bug there is that `AppInit` can return true without initializing the shutdown pipe, which makes `WaitForShutdown` assertion fail (cannot call `WaitForShutdown` without initializing the pipe). The user could have provided the `-help` or `-version` arguments, [see](https://github.com/bitcoin/bitcoin/blob/a36134fcc7b40671d538931f621c8c15ffacc3d2/src/bitcoind.cpp#L142), which prints information and return early without initializing the node.

Overall, I like where those changes are going, but they require to refactor `AppInit` further so the args parsing step, and the execution of commands that return early without initializing the node, are performed in a different function.
If you feel strong on adding it here, I could add it. Not really a thing to introduce more changes for me.

> Maybe drop the "move 'AbortNode' after 'StartShutdown'" commit if it is not needed anymore. Dropping it would help git blame work better and make history more legible. Also https://github.com/bitcoin/bitcoin/pull/27711 is going to remove AbortNode so moving it now will cause an uglier conflict.

k good, didn't know about 27711 removing it. Will drop it.

furszy committed:
refactor: decouple early return commands from AppInit

Cleaned up the init flow to make it more obvious when
the 'exit_status' value will and won't be returned.

This is because it was confusing that `AppInit` was
returning true under two different circumstances:

1) When bitcoind was launched only to retrieve the "-help"
or "-version" information. In this case, the app was
not initialized.

2) When the user triggers a shutdown. In this case,
the app was fully initialized.

furszy commented:
Have quickly drafted the needed changes to make your diff work so we are in the same page, check [61fe1af](https://github.com/bitcoin/bitcoin/pull/27708/commits/61fe1af01418d8b93d38aba8aeb65aecfa6a95cd).

ryanofsky commented on src/init.cpp:
```diff
@@ -800,7 +800,7 @@ std::set<BlockFilterType> g_enabled_filter_types;
 };
 
-bool AppInitBasicSetup(const ArgsManager& args)
+bool AppInitBasicSetup(const ArgsManager& args, NodeContext& node)
```
In commit "refactor: index: use `AbortNode` in fatal error helper" (3c06926cf21dcca3074ef51506f556b2286c299b)

It would be slightly better to just pass a reference to node.exit_status instead of passing the whole NodeContext, because passing NodeContext exposes a lot of other state this function doesn't need to have access to. But it's not important because #27711 gets rid of the InitShutdownState function and will drop the parameter anyway.

furszy replied:
done. Refactored to pass `exit_status` ref.

ryanofsky commented on src/qt/bitcoin.cpp:
```diff
@@ -398,8 +399,7 @@ void BitcoinApplication::initializeResult(bool success, interfaces::BlockAndHead
     qDebug() << __func__ << ": Initialization result: " << success;
 
     // Set exit result.
```
In commit "gui: return EXIT_FAILURE on post-init fatal errors" (c8806e972f2be9f8a05bf11b1f6a8a2e70e01962)

Should delete the "Set exit result" comment here or move it down below. Probably fine to delete because `exit_status.store(EXIT_FAILURE)` should pretty clear by itself.

furszy replied:
done, deleted

ryanofsky commented on src/qt/bitcoin.cpp:
```diff
@@ -464,6 +465,10 @@ void BitcoinApplication::handleNonFatalException(const QString& message)
 }
 
+int BitcoinApplication::getExitStatus() const {
+    return node().context()->exit_status.load();
```
In commit "gui: return EXIT_FAILURE on post-init fatal errors" (c8806e972f2be9f8a05bf11b1f6a8a2e70e01962)

This is ok for now, but if you want to future-proof the code and make it compatible with #10102, it would be better to add an `interfaces::Node::getExitStatus()` method and have GUI call `node().getExitStatus()`. The problem is that with  #10102, `node().context()` will return null when the node is running in a different process than the GUI.

I also think it might be clearer just to delete the `BitcoinApplication::getExitStatus` method entirely and just `return node().getExitStatus()` at the end of `GuiMain()`

furszy replied:
> I also think it might be clearer just to delete the BitcoinApplication::getExitStatus method entirely and just return node().getExitStatus() at the end of GuiMain()

That was my first idea too but `GuiMain()` is a static function with no direct access to the node interface. The interface is created inside the `BitcoinApplication::createNode` class method.

> This is ok for now, but if you want to future-proof the code and make it compatible with https://github.com/bitcoin/bitcoin/pull/10102, it would be better to add an interfaces::Node::getExitStatus() method and have GUI call node().getExitStatus().

Yeah, good. Added `getExitStatus()` to the node interface.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/27708#discussion_r1224765526

> That was my first idea too but `GuiMain()` is a static function with no direct access to the node interface.

I think it can just do `return app.node().getExitStatus();`. The `app.node()` method is public and already used in `GuiMain()`

furszy replied:
Done, added

ryanofsky commented on src/bitcoind.cpp:
```diff
@@ -153,17 +172,6 @@ static bool AppInit(NodeContext& node, int argc, char* argv[])
     std::any context{&node};
     try
     {
-        if (auto error = common::InitConfig(args)) {
```
In commit "refactor: decouple early return commands from AppInit" (61fe1af01418d8b93d38aba8aeb65aecfa6a95cd)

Note for reviewers: Moving the `common::InitConfig` call out of the try block means if InitConfig throws an exception it won't be caught. But this is good and should not change behavior because InitConfig is already catching `std::exception` internally and should just be doing basic parsing and not triggering other types of exceptions.

furszy replied:
yeah, same note applies for the other loop that was moved out of the try catch. The one that calls `IsSwitchChar()`.

ryanofsky reviewed (approved):
Code review ACK 61fe1af01418d8b93d38aba8aeb65aecfa6a95cd. This looks very good. 

Appreciate you cleaning up the code to make it more obvious when the new `exit_status` value will and won't be returned. (And sorry I posted another buggy diff that forgot about the early returns!) Left a few more minor suggestions that you could take or leave.

furszy commented on src/bitcoind.cpp:
```diff
@@ -112,12 +112,8 @@ int fork_daemon(bool nochdir, bool noclose, TokenPipeEnd& endpoint)
 #endif
 
-static bool AppInit(NodeContext& node, int argc, char* argv[])
+static bool ParseArgs(NodeContext& node, int argc, char* argv[])
```
done

furszy commented:
Thanks @ryanofsky!
Feedback tackled. Small [diff](https://github.com/bitcoin/bitcoin/compare/61fe1af01418d8b93d38aba8aeb65aecfa6a95cd..da60d86bfbe7fe49ce7181b5b394c0e3d8c53854).

ryanofsky commented on src/qt/bitcoin.cpp:
```diff
@@ -441,6 +440,7 @@ void BitcoinApplication::initializeResult(bool success, interfaces::BlockAndHead
 #endif
         pollShutdownTimer->start(SHUTDOWN_POLLING_DELAY);
     } else {
+        m_node->context()->exit_status.store(EXIT_FAILURE);
```
In commit "gui: return EXIT_FAILURE on post-init fatal errors" (624340f9acbdd3c2d54c3824248e2c0cff4cf487)

Not necessary to address here, but with #10102 when the node and GUI are running in separate processes, `m_node->context()` will be null, so this will crash.

You could fix and future-proof this by adding a `interfaces::Node::setExitStatus()` function, or just update the existing [`appInitMain()`](https://github.com/bitcoin/bitcoin/blob/153a6882f42fff3fdc63bf770d4c86a62c46c448/src/node/interfaces.cpp#L106-L109) function to set the error code:

```c++
    bool appInitMain(interfaces::BlockAndHeaderTipInfo* tip_info) override
    {
        if (!AppInitMain(*m_context, tip_info)) {
            m_context->exit_status.store(EXIT_FAILURE);
            return false;
        }
        return true;
    }
```

If you take let `appInitMain` set the error code, you could also simplify GUI code even more, skipping `Q_EMIT initializeResult(rv, tip_info)` when `appInitMain` fails and getting rid of the `rv` parameter.

furszy replied:
> You could fix and future-proof this by adding a interfaces::Node::setExitStatus() function, or just update the existing [appInitMain()](https://github.com/bitcoin/bitcoin/blob/153a6882f42fff3fdc63bf770d4c86a62c46c448/src/node/interfaces.cpp#L106-L109) function to set the error code.

Great, added it. Thanks!

> If you let appInitMain set the error code, you could also simplify GUI code even more, skipping Q_EMIT initializeResult(rv, tip_info) when appInitMain fails and getting rid of the rv parameter.

Hmm, that would be nice but, in a quick glance, I don't think that we can remove the `rv` parameter without introducing another signal to the `InitExecutor` class (which wouldn't be bad). Because we still need to call to the shutdown function at the GUI level: the `BitcoinApplication::requestShutdown` that is called inside `BitcoinApplication::initializeResult` when the rv parameter is false.

Still, even when I love to keep cleaning stuff, let's leave it for a follow-up so the gui related changes are minimal here.

ryanofsky reviewed (approved):
Code review ACK da60d86bfbe7fe49ce7181b5b394c0e3d8c53854

Thanks for following up on earlier suggestions. Left some new comments below, but this also looks good in it's current form.

furszy commented:
Updated per feedback, thanks ryanofsky. [Tiny diff](https://github.com/bitcoin/bitcoin/compare/da60d86bfbe7fe49ce7181b5b394c0e3d8c53854..61c569ab6069d04079a0831468eb713983919636):

- Removed `BitcoinApplication::getExitStatus`.
- Moved `exit_status.store(EXIT_FAILURE)` to the interface `appInitMain()` method to be compatible with [#10102](https://github.com/bitcoin/bitcoin/pull/10102).

pinheadmz commented:
concept ACK

I was looking for this writing https://github.com/bitcoin/bitcoin/pull/27850 was surprised that a fatal internal error didn't satisfy the python test for assert init error on startup

sedited reviewed (approved):
ACK 61c569ab6069d04079a0831468eb713983919636

ryanofsky reviewed (approved):
Code review ACK 61c569ab6069d04079a0831468eb713983919636

This should be a really nice change because it actually makes `bitcoind` return an error code when there is an error. But it also clears out `StartShutdown` clutter and makes the `main()` function comprehensible, so is a nice code improvment.

I think this would be ready for merge if another reviewer added an ACK.

pinheadmz commented on src/shutdown.cpp:
```diff
@@ -28,6 +31,7 @@ bool AbortNode(const std::string& strMessage, bilingual_str user_message)
         user_message = _("A fatal internal error occurred, see debug.log for details");
     }
     InitError(user_message);
+    Assert(g_exit_status)->store(EXIT_FAILURE);
```
Does this make `InitShutdownState()` required before calling `AbortNode()` ? 
I suppose that could be added to the comment here. I'm just making sure I understand, I don't think you need to retouch:

https://github.com/bitcoin/bitcoin/blob/6f5f37eefd425faabd9a97a3c3028395c34f08c4/src/shutdown.h#L14-L17

pinheadmz commented on src/shutdown.cpp:
```diff
@@ -20,6 +21,8 @@
 #include <condition_variable>
 #endif
 
+static std::atomic<int>* g_exit_status{nullptr};
```
I am guessing we need a new global here because we can't otherwise access `NodeContext::exit_status`?

TheCharlatan replied:
FWIW, this could be de-globalized again by the interrupt class introduced in https://github.com/bitcoin/bitcoin/pull/27861.

pinheadmz reviewed (approved):
ACK 61c569ab6069d04079a0831468eb713983919636

A few non-breaking questions below. Reviewed each commit and ran all tests locally. Confirmed that #2039 behavior now exits with code `1` instead of `0` like it did on master. Also confirmed that remaining `StartShutdown()` calls are all user-triggered.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 61c569ab6069d04079a0831468eb713983919636
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmSHODIACgkQ5+KYS2KJ
yTpd4g/+I56MeCxiLqHXVtFe2yUw9JY1wYNeo6kuLU2A5Yx+kWeTBKZl9dA/lTIu
5/TYbFaTai5px1HZzynnfD+XrFLtjaxynGqHhbfRmX8cGUyo5WkRRY6IcYUuqiXs
FBUsTqlXrEq00j012RL63xpiZhswUZTEpw3BzHD5c1IlK7FkdCWCVT0xl3o68ivs
W0pfXRXnAcAtseWgMlrZtPHGaEsQS1CMHAaz2KXvRPbOCu0p0vjAuGkzqpBZuu3x
2kzVcPnuUunb6F+t62Q2Mrk4bCuKYlT9I5CozGVV8W9rWJODvzKyTCmTLCXbIxC0
MrGN6DHn0AzBbCBm6E65o5B80czNzM0Wci/7ZElmjBkDJxWP0KPDtaNp9QsazYbO
AVWhKfMMBDEqmwRyXvMK1iv/zJrZ2Y2QQf/LPt1vbLjVwHT2wF7HyEOUJHKKmrI2
fLWf62JKvBvrrNdeb12s5ZaMwMgcM5UA/8H2YuDw/zOvgJYZZiXAiRI3ETDvsm5B
dFIgn0M5njMq3UFgRl2hHW/qBwKiufs8QvrKw8zxuXxF5H+4sQcwCVtiwe//dsDY
179WwJQcQ3t7iwg1TNgscv9nVY5nRJvPzHs4iopLbPhks60akDic1g6ruMCdbAEj
/Md+oVk7d59HJLvhDlR6xpS8iKycyadFcsiMlOuK/Xn56ZPLKt0=
=N/HN
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

theStack commented on test/functional/test_framework/test_node.py:
```diff
@@ -377,17 +377,23 @@ def is_node_stopped(self):
+        # unless 'expected_ret_code' is provided.
+        if expected_ret_code is not None:
+            assert return_code == expected_ret_code, self._node_msg(
+                "Node returned unexpected exit code (%d) vs (%d) when stopping" % (return_code, expected_ret_code))
```
nit, could take the opportunity to change to f-strings here (and below), as [suggested in the functional test style guidelines](https://github.com/bitcoin/bitcoin/blob/361a0c00b3fdac7149c027c2c69e4db60ee5aa86/test/functional/README.md?plain=1#L39):
```suggestion
                 f"Node returned unexpected exit code ({return_code}) vs ({expected_ret_code}) when stopping")
```

theStack reviewed (approved):
Code-review ACK 61c569ab6069d04079a0831468eb713983919636

maflcko reviewed (commented):
left a question / test nit for follow-up
